### PR TITLE
drivers: mbox: add initial drivers for nRF VEVIF & BELLBOARD

### PR DESCRIPTION
**Pull request by gmarull**
Tested with:

```
west build -b nrf54h20pdk/nrf54h20/cpuapp samples/drivers/mbox -T sample.drivers.mbox.nrf54h20_vevif -p
west build -b nrf54h20pdk/nrf54h20/cpuapp samples/drivers/mbox -T sample.drivers.mbox.nrf54h20_bellboard -p
```

Pending:
- DTize mbox API, so that things are more ergonomic (dt_spec from #mbox-cells) see https://github.com/zephyrproject-rtos/zephyr/pull/69390
- Refactor sample to use dt-spec stuff (depends on https://github.com/zephyrproject-rtos/zephyr/pull/69390)
- Test VEVIF+BELLBOARD together (ie bidirectional) (depends on https://github.com/zephyrproject-rtos/zephyr/pull/69390, as current sample doesn't deal with 2 different mboxes now)

### CODE REVIEW

**Review by masz-nordic — APPROVED** [review]

File: samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+mbox: &cpuapp_bellboard {
+	interrupts = <96 NRF_DEFAULT_IRQ_PRIORITY>;
+	interrupt-names = "bellboard0";
+	status = "okay";
+};
```
Comment (57300) [inline samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay:10]: These might be worth moving to the default board configuration, sans label.

File: samples/drivers/mbox/sample.yaml [hunk 2]
```diff
@@ -23,6 +23,47 @@ tests:
...
         - "Pong \\(on channel 1\\)"
+
+  sample.drivers.mbox.nrf54h20_vevif:
+    platform_allow:
+      - nrf54h20pdk_nrf54h20_cpuapp
+    integration_platforms:
+      - nrf54h20pdk_nrf54h20_cpuapp
+    extra_args:
+      mbox_SNIPPET=nordic-ppr
+      mbox_OVERLAY_CONFIG=boards/nrf54h20pdk_nrf54h20_cpuapp_vevif.conf
```
Comment (57300) [inline samples/drivers/mbox/sample.yaml:34]: `OVERLAY_CONFIG` is deprecated.
```suggestion
      mbox_EXTRA_CONF_FILE=boards/nrf54h20pdk_nrf54h20_cpuapp_vevif.conf
```

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	.register_callback = bellboard_local_register_callback,
+	.set_enabled = bellboard_local_set_enabled,
+};
+
+#define BELLBOARD_IRQ_CONNECT(grp, grp_idx)                                                        \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, grp),                                                  \
+		    (static const uint8_t idx_##grp = grp_idx;                                     \
+		     IRQ_CONNECT(DT_INST_IRQ_BY_NAME(0, grp, irq),                                 \
+				 DT_INST_IRQ_BY_NAME(0, grp, priority), bellboard_local_isr,       \
+				 &idx_##grp, 0);),                                                 \
```
Comment (57300) [inline drivers/mbox/mbox_nrf_bellboard_local.c:172]: Why does the second to last param need to be a pointer to a byte in ROM?

**Review by 57300 — COMMENTED** [review]
LGTM, just minor comments below.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_bellboard_local
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/mbox.h>
+
+#include <haly/nrfy_bellboard.h>
+
+#define BELLBOARD_GROUPS 4U
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: I couldn't find this definition, is it in a new HAL version? Anyway, used such name for the local define now, so it can be updated later easily.

File: samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay (hunk 1 above)
Comment (gmarull) [reply samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay]: Could you suggest a default IRQ configuration?

File: samples/drivers/mbox/sample.yaml (hunk 2 above)
Comment (gmarull) [reply samples/drivers/mbox/sample.yaml]: done

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 3 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: indeed, just passed number as address.

**Review by carlocaione — DISMISSED** [review]

File: samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay (hunk 1 above)
Comment (57300) [reply samples/drivers/mbox/boards/nrf54h20pdk_nrf54h20_cpuapp_bellboard.overlay]: Let's come back to it later.

**Review by 57300 — DISMISSED** [review]

**Review by anangl — CHANGES_REQUESTED** [review]

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+	ARG_UNUSED(dev);
+
+	for (uint8_t i = 0U; i < ARRAY_SIZE(grps); i++) {
+		if (grps[i] != NULL) {
+			max_channels += NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT;
+		}
+	}
+
+	return max_channels;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c:115]: indeed, this is wrong, will fix it

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+};
+
+#define BELLBOARD_GRP_MAYBE_DEFINE(grp)                                                            \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, grp),                                                  \
+		    (static struct mbox_bellboard_local_group g_##grp;), ())
+
+#define BELLBOARD_GRP_GET_OR_NULL(grp) COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, grp), (&g_##grp), (NULL))
+
+BELLBOARD_GRP_MAYBE_DEFINE(bellboard0)
+BELLBOARD_GRP_MAYBE_DEFINE(bellboard1)
```
Comment (anangl) [inline drivers/mbox/mbox_nrf_bellboard_local.c:29]: This is a bit confusing as `bellboardX` looks like an instance of the BELLBOARD peripheral. I'd suggest to use `irqX` instead.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+#include <zephyr/drivers/mbox.h>
+
+#include <haly/nrfy_bellboard.h>
+
+#define BELLBOARD_IRQ_COUNT 4U
+
+/* BELLBOARD runtime resources for each group */
+struct mbox_bellboard_local_group {
+	mbox_callback_t cb[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
+	void *user_data[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
```
Comment (anangl) [inline drivers/mbox/mbox_nrf_bellboard_local.c:18]: There is no need to use separate arrays of callbacks and user data pointers for particular groups. Just one instance of each is enough.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+static int bellboard_local_set_enabled(const struct device *dev, uint32_t id, bool enable)
+{
+	struct mbox_bellboard_local_group *grp;
+	uint8_t grp_idx;
+	int ret;
+
+	ARG_UNUSED(dev);
+
+	ret = bellboard_local_grp_get(id, &grp_idx);
```
Comment (anangl) [inline drivers/mbox/mbox_nrf_bellboard_local.c:148]: This finds only the first group to which a given event is assigned but it is technically possible that one event is assigned to several groups. It would be good to handle this possibility.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_bellboard_local
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/mbox.h>
+
+#include <haly/nrfy_bellboard.h>
+
+#define BELLBOARD_IRQ_COUNT 4U
```
Comment (anangl) [inline drivers/mbox/mbox_nrf_bellboard_local.c:13]: This number can potentially differ between cores in future devices. Perhaps it should be defined in devicetree.

File: drivers/mbox/Kconfig.nrf_bellboard [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config MBOX_NRF_BELLBOARD_LOCAL
+	bool "nRF BELLBOARD local driver"
+	depends on DT_HAS_NORDIC_NRF_BELLBOARD_LOCAL_ENABLED
+	default y
+
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/mbox/Kconfig.nrf_bellboard]: fixed

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 7 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: fixed, good catch

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 8 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: reworked driver to support multi-irq mapping

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 9 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: that's true, I guess there are 2 options here:

- Add `nordic,interrupt-lines = <N>;` in DT
- Always force `interrupts` to define the N lines, and add `nordic,interrupt-mask` which would indicate enabled lins (eg 0x3 = irq0 and irq1), this would require reworking the driver a bit, and would change how bellboard is used in DT, so it's a good time to decide.

The (2) looks more user-friendly at a first glance, but, AFAIK, the idea of using multiple IRQs is to play with priorities, in which case... one needs to re-define the `interrupts` property entirely. So (1) seems a better approach.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+/* BELLBOARD IRQ mappings */
+#define BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(idx)                                                     \
+	COND_CODE_1(DT_INST_PROP_HAS_IDX(0, nordic_interrupt_mapping, idx),                        \
+		    (DT_INST_PROP_BY_IDX(0, nordic_interrupt_mapping, idx)), (0U))
+
+static const uint32_t irq_mappings[BELLBOARD_NUM_IRQS] = {
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(0),
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(2),
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(4),
+	BELLBOARD_GET_IRQ_MAPPING_OR_ZERO(6),
```
Comment (anangl) [inline drivers/mbox/mbox_nrf_bellboard_local.c:41]: I think it's not going to work as intended. For instance, if you use the example from binding:
```
      interrupts = <98 NRF_DEFAULT_IRQ_PRIORITY>,
                   <99 NRF_DEFAULT_IRQ_PRIORITY>;
      interrupt-names = "irq2", "irq3";
      nordic,interrupt-mapping = <0x0000000f 2>, <0x000000f0 3>;
```
you'll get: `irqs = { 0, 0, 98, 99 };` and `irq_mappings = { 0x0000000f, 0x000000f0, 0, 0 };`

Comment (anangl) [reply drivers/mbox/mbox_nrf_bellboard_local.c:41]: How about using designated initializers and `LISTIFY` to fill `irq_mappings`?

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	     "# interrupt exceeds maximum");
+
+BUILD_ASSERT(DT_INST_PROP_LEN(0, nordic_interrupt_mapping) == 2 * DT_NUM_IRQS(DT_DRV_INST(0)),
+	     "# interrupt mappings != # interrupts");
+
+/* BELLBOARD IRQ list */
+#define BELLBOARD_IRQ_GET_OR_ZERO(irq_)                                                            \
+	COND_CODE_1(DT_INST_IRQ_HAS_NAME(0, irq_), (DT_INST_IRQ_BY_NAME(0, irq_, irq)), (0U))
+
+static const uint8_t irqs[BELLBOARD_NUM_IRQS] = {
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: no longer needed in current revision

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+static NRF_BELLBOARD_Type *bellboard = (NRF_BELLBOARD_Type *)DT_INST_REG_ADDR(0);
+
+/* BELLBOARD runtime resources */
+static mbox_callback_t cbs[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
+static void *cbs_ctx[NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT];
+static uint32_t enabled_masks[BELLBOARD_NUM_IRQS];
+
+static void bellboard_local_isr(const void *parameter)
+{
+	uint8_t irq_idx = (uintptr_t)parameter;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: fixed

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+			if ((*enabled_mask & BIT(id)) == 0U) {
+				return -EALREADY;
+			}
+
+			*enabled_mask &= ~BIT(id);
+			nrfy_bellboard_int_disable(bellboard, i, BIT(id));
+
+			if (*enabled_mask == 0U) {
+				irq_disable(irqs[i]);
+			}
```
[the remark this answers is not in the document]
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: irqs are now always enabled, and int_enable/disable is only used

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 11 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: good catch, please check current revision

Comment (anangl) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: > good catch, please check current revision

It still won't handle properly the above example. I think something like this should be used:
```
#define EVT_MAPPING_ITEM(idx) DT_INST_PROP_BY_IDX(0, nordic_interrupt_mapping, idx)
#define BELLBOARD_GET_EVT_MAPPING(idx, _)				\
	COND_CODE_1(DT_INST_PROP_HAS_IDX(0, nordic_interrupt_mapping,	\
					 UTIL_INC(UTIL_X2(idx))),	\
		([EVT_MAPPING_ITEM(UTIL_INC(UTIL_X2(idx)))] =		\
			EVT_MAPPING_ITEM(UTIL_X2(idx)),),		\
		())

static const uint32_t evt_mappings[BELLBOARD_NUM_IRQS] = {
	LISTIFY(DT_NUM_IRQS(DT_DRV_INST(0)), BELLBOARD_GET_EVT_MAPPING, ())
};
```
Then it won't be also needed to require that "# interrupt mappings == # interrupts";

Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c]: you're right, updated the driver to user you sample, thanks

File: dts/bindings/mbox/nordic,nrf-bellboard-local.yaml [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+  nordic,interrupt-mapping:
+    type: array
+    required: true
+    description: |
+      Set of interrupt mapping pairs. Each pair consists of a bitmask and an
+      interrupt identifier. The bitmask is used to indicate which of the 32
+      possible events are mapped to the given interrupt. For example, given
+      <0x0000000f 2>, the first four events are mapped to interrupt 2
+      (irq2). There needs to be at least one mapping for each interrupt.
```
Comment (anangl) [inline dts/bindings/mbox/nordic,nrf-bellboard-local.yaml:43]: This last sentence can be dropped now.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+#include <zephyr/sys/util.h>
+
+#include <hal/nrf_bellboard.h>
+
+#define BELLBOARD_NUM_IRQS 4U
+
+BUILD_ASSERT(DT_NUM_IRQS(DT_DRV_INST(0)) <= BELLBOARD_NUM_IRQS, "# interrupt exceeds maximum");
+
+BUILD_ASSERT(DT_INST_PROP_LEN(0, nordic_interrupt_mapping) == 2 * DT_NUM_IRQS(DT_DRV_INST(0)),
+	     "# interrupt mappings != # interrupts");
```
Comment (anangl) [inline drivers/mbox/mbox_nrf_bellboard_local.c:19]: This is also not needed. Just checking if the length is a multiple of 2 should be sufficient.

**Review by anangl — DISMISSED** [review]

**Review by anangl — DISMISSED** [review]

**Review by anangl — DISMISSED** [review]

**Review by 57300 — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

**Review by 57300 — DISMISSED** [review]

**Review by anangl — DISMISSED** [review]

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts [hunk 17]
```diff
@@ -93,6 +96,32 @@
...
 };
 
+&cpuapp_bellboard {
+	interrupts = <96 NRF_DEFAULT_IRQ_PRIORITY>;
+	interrupt-names = "irq0";
+	/* irq0: 0: cpuapp-cpusec, 13: cpuapp-cpuppr, 18: cpuapp-cpurad */
+	nordic,interrupt-mapping = <0x00042001 0>;
+};
+
+&cpusec_cpuapp_ipc {
```
Comment (hubertmis) [inline boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:106]: I understand the reason to avoid repeating code and reusing parts of this node between both peers in communication, but the result is a hard-to-understand content of this node because it is scattered across multiple files. It's hard to follow which mbox instance is tx and which is rx.

File: drivers/mbox/mbox_nrf_vevif_local.c [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+		cbs.enabled_mask |= BIT(id);
+		irq_enable(vevif_irqs[id]);
+	} else {
+		if ((cbs.enabled_mask & BIT(id)) == 0U) {
+			return -EALREADY;
+		}
+
+		cbs.enabled_mask &= ~BIT(id);
+		irq_disable(vevif_irqs[id]);
```
Comment (hubertmis) [inline drivers/mbox/mbox_nrf_vevif_local.c:96]: I would expect a condition here to check if the event being disabled is the last event assigned to this IRQ, before the IRQ gets disabled.
Without this condition, there is a risk that calling this function for event 13 while events 12, 13, and 18 are assigned to IRQ 0, the event 13 gets disabled and a side-effect it disables IRQ line for still active events 12 and 18

File: samples/drivers/mbox/src/main.c [hunk 19]
```diff
@@ -19,24 +18,25 @@ static void callback(const struct device *dev, uint32_t channel,
 
 int main(void)
 {
-	const struct device *dev;
+	int ret;
+	const struct device *const dev = DEVICE_DT_GET(DT_NODELABEL(mbox));
 
 	printk("Hello from APP\n");
```
Comment (hubertmis) [inline samples/drivers/mbox/src/main.c:24]: You could consider renaming this to something like MAIN or HOST. This CPU does not have to do anything with an application.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+	ARG_UNUSED(dev);
+
+	nrf_bellboard_int_disable(bellboard, 0, evt_mappings[0]);
+	nrf_bellboard_int_disable(bellboard, 1, evt_mappings[1]);
+	nrf_bellboard_int_disable(bellboard, 2, evt_mappings[2]);
+	nrf_bellboard_int_disable(bellboard, 3, evt_mappings[3]);
+
+	for (uint8_t i = 0U; i < NRF_BELLBOARD_EVENTS_TRIGGERED_COUNT; i++) {
+		if ((evt_all_mappings & BIT(i)) != 0U) {
+			nrf_bellboard_event_clear(bellboard, nrf_bellboard_triggered_event_get(i));
```
Comment (hubertmis) [inline drivers/mbox/mbox_nrf_bellboard_local.c:156]: I don't think the driver should clear all the events on initialization.
We can imagine that a CPU X is in a SoftOff state and it is woken up by a BELLBOARD event by another CPU Y. The event is pending. Then the CPU X runs a complete cold-boot procedure, including the bellboard driver initialization. After the initialization is complete and the application code enables bellboard events and IRQs, the pending event must be handled. If the event gets cleared during the driver initialization, the event will be lost and CPU X won't have any idea why it was woken up.

File: drivers/mbox/mbox_nrf_bellboard_local.c [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+			*evt_enabled_mask |= BIT(id);
+			nrf_bellboard_int_enable(bellboard, i, BIT(id));
+		} else {
+			if ((*evt_enabled_mask & BIT(id)) == 0U) {
+				return -EALREADY;
+			}
+
+			*evt_enabled_mask &= ~BIT(id);
+			nrf_bellboard_int_disable(bellboard, i, BIT(id));
```
Comment (hubertmis) [inline drivers/mbox/mbox_nrf_bellboard_local.c:117]: Similarly to VEVIF, I don't think interrupts should get disabled unconditionally, because other still active channels may be assigned to the same interrupt.

Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c:117]: Note that here we’re never disabling IRQ, but changing the interrupt enabled masks.

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 20 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c:156]: Will drop this. Note that I inherited this code from nrfy, so it has always been done like that. I guess it wasn’t the right thing to do.

File: samples/drivers/mbox/src/main.c (hunk 19 above)
Comment (gmarull) [reply samples/drivers/mbox/src/main.c:24]: Sample needs a bit of love, I’ll rework it as part of the https://github.com/zephyrproject-rtos/zephyr/pull/69390 when I rebase it

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts (hunk 17 above)
Comment (gmarull) [reply boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:106]: @57300

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts [hunk 22]
```diff
@@ -8,8 +8,11 @@
 
 #include <nordic/nrf54h20_enga_cpuapp.dtsi>
 #include "nrf54h20pdk_nrf54h20-memory_map.dtsi"
+#include "nrf54h20pdk_nrf54h20-ipc_conf.dtsi"
 #include "nrf54h20pdk_nrf54h20-pinctrl.dtsi"
 
+/delete-node/ &cpusec_cpurad_ipc;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:14]: I think deleting is preferred, because this actually doesn’t exist for this core. This could be avoided by not creating any “common” file I guess.

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi [hunk 23]
```diff
@@ -15,7 +105,7 @@
 			perm-execute;
 		};
 
-		ram3x_dma_region: memory@2fc07000 {
+		shared_ram3x_region: memory@2fc07000 {
 			compatible = "nordic,owned-memory";
 			reg = <0x2fc07000 DT_SIZE_K(4)>;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:110]: @57300

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi [hunk 24]
```diff
@@ -6,6 +6,96 @@
...
+			};
+
+			cpurad_cpuapp_ipc_shm: memory@800 {
+				reg = <0x800 DT_SIZE_K(2)>;
+			};
+		};
+
+		shared_ram20_region: memory@2f88f000 {
+			compatible = "nordic,owned-memory";
+			reg = <0x2f88f000 DT_SIZE_K(4)>;
```
[the remark this answers is not in the document]
Comment (gmarull) [reply boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:74]: Note this is all “soc1” or EngA

File: dts/common/nordic/nrf54h20_enga.dtsi [hunk 25]
```diff
@@ -281,6 +306,15 @@
...
+				cpuppr_vevif_remote: mailbox@0 {
+					compatible = "nordic,nrf-vevif-remote";
+					reg = <0x0 0x1000>;
+					status = "disabled";
+					#mbox-cells = <1>;
+					nordic,tasks = <16>;
+					nordic,tasks-mask = <0xfffffff0>;
+				};
+
 				cpuppr_clic: interrupt-controller@1000 {
```
[the remark this answers is not in the document]
Comment (57300) [reply dts/common/nordic/nrf54h20_enga.dtsi:336]: This discussion can continue in #69281

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi (hunk 23 above)
[the remark this answers is not in the document]
Comment (57300) [reply boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:110]: This is not Final Product, so the 4 KiB boundary applies. @gmarull Up to you if you want to add the `TODO`.

File: boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts (hunk 17 above)
Comment (57300) [reply boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:106]: I was betting on the mbox placement being easy to deduce from the IPC node label. If it's `cpuxxx_cpuyyy_ipc`, then the first one is `cpuxxx_bellboard/vevif` and the second is `cpuyyy_bellboard/vevif`. If you're building for `cpuxxx`, then the first one is RX and the second is TX (to `cpuyyy`), and vice versa for the other core.

In addition, it should also be possible to tuck away the shared memory configuration in the common file, if we deprecate the `tx/rx-region` properties in favor of `memory-regions` and `memory-region-names`.

This information is not scattered just for the sake of avoiding repetition. It makes sense to track channel and memory configuration at system level, as opposed to individual core level.

But if that's too confusing, then I don't mind going back to duplicating nodes.

File: samples/drivers/mbox/remote/CMakeLists.txt [hunk 26]
```diff
@@ -15,7 +15,8 @@ if(("${BOARD}" STREQUAL "nrf5340dk_nrf5340_cpunet") OR
   ("${BOARD}" STREQUAL "mimxrt1170_evk_cm4") OR
   ("${BOARD}" STREQUAL "mimxrt1160_evk_cm4") OR
   ("${BOARD}" STREQUAL "lpcxpresso55s69_cpu1") OR
-  ("${BOARD}" STREQUAL "adp_xc7k_ae350"))
+  ("${BOARD}" STREQUAL "adp_xc7k_ae350") OR
+  ("${BOARD}" STREQUAL "nrf54h20pdk_nrf54h20_cpuppr"))
```
Comment (nordicjm) [inline samples/drivers/mbox/remote/CMakeLists.txt:19]: Needs converting to hwmv2

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: samples/drivers/mbox/remote/CMakeLists.txt (hunk 26 above)
Comment (gmarull) [reply samples/drivers/mbox/remote/CMakeLists.txt]: done

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 20 above)
Comment (gmarull) [reply drivers/mbox/mbox_nrf_bellboard_local.c:156]: Note: I'll open a follow-up PR resolving any remaining issues after we have a call to discuss more details, since the logic in this PR has been largely based on what we already had before. Also, it's already part of sdk-nrf in its current form (and it works "well enough" to unblock certain work), so we do not have the same pressure as before to discuss technical aspects as before :-)

**Review by nordicjm — DISMISSED** [review]

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 20 above)
Comment (nika-nordic) [reply drivers/mbox/mbox_nrf_bellboard_local.c:156]: What I am afraid of is that scheme of clearing events in nrfx drivers and nrfy layer during initialization is established for quite some time. Rationale is that during initialization driver applies clear & known hardware configuration, to avoid situation in which leftover peripheral register contents alter the driver behavior.

If we decide to change the nrfx behavior in this manner (that is to handle events that might be triggered before driver initialization and shall be considered as valid) IMO such change should be done systematically across all drivers, instead of doing it ad-hoc here

File: drivers/mbox/mbox_nrf_vevif_local.c (hunk 18 above)
Comment (masz-nordic) [reply drivers/mbox/mbox_nrf_vevif_local.c:96]: I think this is unnecessary at this point. VEVIF has separate IRQ for every event.

**Comment by gmarull** [comment]
@nordicjm @anangl @57300 ping

**Review by nordicjm — APPROVED** [review]

**Review by anangl — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

File: drivers/mbox/mbox_nrf_bellboard_local.c (hunk 20 above)
Comment (hubertmis) [reply drivers/mbox/mbox_nrf_bellboard_local.c:156]: IPC signaling peripherals might be a special case requiring a different approach. But let's discuss it in the "follow up PR" scope mentioned by @gmarull

File: drivers/mbox/mbox_nrf_vevif_local.c (hunk 18 above)
Comment (hubertmis) [reply drivers/mbox/mbox_nrf_vevif_local.c:96]: Nice! I thought they share IRQs like in the BELLBOARD
